src/ast/str_util.rs

```diff
@@ -0,0 +1,306 @@
+                None => break,
+                Some('\'') => match input.peek() {
+                    Some('\'') => {
+                        input.next().unwrap();
+                        output.push_str("''");
+                    }
+                    Some('$') => {
+                        input.next().unwrap();
+                        output.push('$');
+                    }
+                    Some('\\') => {
+                        input.next().unwrap();
+                        match input.next() {
+                            None => break,
+                            Some('n') => output.push('\n'),
+                            Some('r') => output.push('\r'),
+                            Some('t') => output.push('\t'),
+                            Some(c) => output.push(c),
+                        }
+                    }
+                    _ => break,
+                },
+                Some(c) => {
+                    output.push('\'');
+                    output.push(c);
+                }
+            },
+            Some(c) => output.push(c),
+        }
+    }
+    output
+}
+
+pub(crate) fn indention(s: &str) -> impl Iterator<Item = char> + '_ {
+    s.chars().take_while(|&c| c != '\n' && c.is_whitespace())
+}
+
+/// Remove common indention in string
+pub fn remove_common_indent(input: &str) -> String {
+    let mut common = std::usize::MAX;
+    for line in input.lines() {
+        let indent = indention(line).count();
+        if line.chars().count() == indent {
+            // line is empty, ignore indention
+            continue;
+        }
+        common = common.min(indent);
+    }
+
+    remove_indent(input, true, common)
+}
+/// Remove a specified max value of indention from each line in a string after
+/// a specified starting point
+pub fn remove_indent(input: &str, initial: bool, indent: usize) -> String {
+    let mut output = String::new();
+    let mut start = 0;
+    if initial {
+        // If the first line is whitespace, ignore it completely
+        let iter = input.chars().take_while(|&c| c != '\n');
+        if iter.clone().all(char::is_whitespace) {
+            start += iter.map(char::len_utf8).sum::<usize>() + /* newline */ 1;
+            if start >= input.len() {
+                // There's nothing after this whitespace line
+                return output;
+            }
+        } else {
+            // Otherwise, skip like normal
+            start += indention(input).take(indent).map(char::len_utf8).sum::<usize>();
+        }
+    }
+    loop {
+        start += indention(&input[start..]).take(indent).map(char::len_utf8).sum::<usize>();
+        let end = input[start..].find('\n').map(|i| start + i + 1);
+        {
+            let end = end.unwrap_or(input.len());
+            output.push_str(&input[start..end]);
+        }
+        start = match end {
+            Some(end) => end,
+            None => break,
+        };
+    }
```

Comment:
It might be a good idea to rewrite this part to leverage `input.lines()` instead.
e.g.
```rust
    input.lines()
        .flat_map(|i| {
            let expanded_indent = indention(i).take(indent).map(char::len_utf8).sum::<usize>();
            &i[expanded_indent..]
        })
        .collect()
```
Also, could we put `it.map(char::len_utf8).sum::<usize>()` into a separate function?